### PR TITLE
'SoundFileFactory' implementation overhaul

### PR DESCRIPTION
**Pull request by vittorioromeo**
tl;dr:

- Use `std::unordered_map` inside `SoundFileFactory` instead of re-implementing a bad version of it using `std::vector`.

- Use Meyers' Singletons for the maps instead of `inline static` variables (likely fixes #2866). This allows cleaner pre-population of the maps with

- Add `isReaderRegistered<T>` and `isWriterRegistered<T>` functions.

- Add tests for registration/deregistration of custom readers/writers.

- Some minor refactoring in the sound example.

### CODE REVIEW

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
Might want to hold off with merging this to prevent yet another rebase of #2749

**Comment by vittorioromeo** [comment]
> Might want to hold off with merging this to prevent yet another rebase of #2749

I volunteer to perform the rebase myself immediately after this is merged :)

**Comment by vittorioromeo** [comment]
BTW, this PR is also blocking #2866, I think it's worthwhile to merge it before #2749.

File: src/SFML/Audio/SoundFileFactory.cpp
Comment (eXpl0it3r) [inline src/SFML/Audio/SoundFileFactory.cpp:144]: I'd personally keep it explicit, so it's clear what we're pulling from the `priv` namespace

Comment (vittorioromeo) [reply src/SFML/Audio/SoundFileFactory.cpp]: Done

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

**Comment by vittorioromeo** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]
Since you volunteered to rebase the miniaudio branch... 😄

**Pull request merged**
